app/questionnaire/navigation.py

```diff
@@ -0,0 +1,151 @@
+from app.questionnaire.location import Location
+from app.questionnaire.rules import evaluate_repeat
+
+logger = logging.getLogger(__name__)
+
+
+class Navigation(object):
+    """
+    Reads navigation config from the schema and returns collections of dicts that describe
+    completion status of each group
+    """
+    def __init__(self, survey_json, metadata=None, answer_store=None, completed_blocks=None):
+        self.survey_json = survey_json
+        self.metadata = metadata or {}
+        self.answer_store = answer_store or AnswerStore()
+        self.completed_blocks = completed_blocks or []
+
+    def build_navigation(self, current_group_id, current_group_instance):
+        """
+        Build navigation based on the current group/instance selected
+
+        :param current_group_id:
+        :param current_group_instance:
+        :return:
+        """
+        if 'navigation' not in self.survey_json:
+            return None
+
+        navigation = []
+
+        for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
+            first_location = Location(group['id'], 0, group['blocks'][0]['id'])
+            is_current_group = group['id'] == current_group_id
+
+            default_highlight = self._should_highlight(group, current_group_id)
+
+            logger.debug("Building frontend navigation for group %s", group)
+
+            repeating_rule = SchemaHelper.get_repeat_rule(group)
+            completed_id = group['completed_id'] if 'completed_id' in group else group['blocks'][-1]['id']
+
+            if repeating_rule:
+                no_of_repeats = evaluate_repeat(repeating_rule, self.answer_store)
+
+                if repeating_rule['type'] == 'answer_count':
+                    link_names = self._generate_link_names(repeating_rule)
+
+                    navigation += self._generate_repeated_items(link_names, first_location, completed_id, no_of_repeats,
+                                                                is_current_group, current_group_instance)
+
+                elif no_of_repeats > 0:
+                    is_completed = self._is_complete(current_group_instance, completed_id)
+                    navigation.append(self._generate_item(group['title'], is_completed, first_location, default_highlight))
+            else:
+                is_completed = self._is_complete(current_group_instance, completed_id)
+                navigation.append(self._generate_item(group['title'], is_completed, first_location, default_highlight))
+
+        return navigation
+
+    @staticmethod
+    def _should_highlight(group, current_group_id):
+        is_current_group = group['id'] == current_group_id
+        current_group_in_highlight = 'highlight_when' in group and current_group_id in group['highlight_when']
+
+        return (current_group_in_highlight and not is_current_group) or is_current_group
+
+    def _is_complete(self, group_instance, completed_id, repeating=False):
```

Comment:
This is a flag to decide whether the item being marked as complete is actually a repeating item. Maybe therefore this should actually be two methods making the check rather than bundling, seeing as how we should be fluent about this and all :)